cpp/src/quantiles/tdigest/tdigest_aggregation.cu

```diff
@@ -1021,6 +1021,90 @@ struct group_key_func {
   }
 };
 
+// merges all the tdigests within each group. returns a table containing 2 columns:
+// the sorted means and weights.
+template <typename GroupOffsetIter>
+std::unique_ptr<table> generate_merged_centroids(tdigest_column_view const& tdv,
+                                                 GroupOffsetIter group_offsets,
+                                                 size_type num_groups,
+                                                 rmm::cuda_stream_view stream)
+{
+  auto temp_mr = rmm::mr::get_current_device_resource();
+
+  auto const total_merged_centroids = tdv.means().size();
+
+  // output table is the merged centroids (means, weights)
+  std::vector<std::unique_ptr<cudf::column>> cols;
+  cols.reserve(2);
+  cols.push_back(std::make_unique<cudf::column>(
+    data_type{type_id::FLOAT64},
+    total_merged_centroids,
+    rmm::device_buffer{sizeof(double) * total_merged_centroids, stream, temp_mr},
+    rmm::device_buffer{0, stream, temp_mr},
+    0));
+  cols.push_back(std::make_unique<cudf::column>(
+    data_type{type_id::FLOAT64},
+    total_merged_centroids,
+    rmm::device_buffer{sizeof(double) * total_merged_centroids, stream, temp_mr},
+    rmm::device_buffer{0, stream, temp_mr},
+    0));
+  auto result = std::make_unique<cudf::table>(std::move(cols));
+
+  // each group represents a collection of tdigest columns. each row is 1 tdigest.
+  // within each group, we want to sort all the centroids within all the tdigests
+  // in that group, using the means as the key. the "outer offsets" represent the indices of the
+  // tdigests, and the "inner offsets" represents the list of centroids for a particular tdigest.
+  //
+  //  rows
+  //  ----        centroid 0 ---------
+  //  tdigest 0   centroid 1
+  //  ----        centroid 2  group 0
+  //  tdigest 1   centroid 3
+  //  ----        centroid 4 ---------
+  //  tdigest 2   centroid 5
+  //  ----        centroid 6  group 1
+  //  tdigest 3   centroid 7
+  //              centroid 8
+  //  ----        centroid 9 --------
+  auto inner_offsets    = tdv.centroids().offsets();
+  auto centroid_offsets = cudf::detail::make_counting_transform_iterator(
+    0,
+    cuda::proclaim_return_type<size_type>(
+      [group_offsets, inner_offsets = tdv.centroids().offsets().begin<size_type>()] __device__(
+        size_type i) { return inner_offsets[group_offsets[i]]; }));
+
+  // perform the sort using the means as the key
+  size_t temp_size;
+  CUDF_CUDA_TRY(
+    cub::DeviceSegmentedSort::SortPairs(nullptr,
```

Comment:
Any idea about the performance of `thrust::sort` comparing to this?

Reply:
Thrust does not have a segmented sort equivalent.

Reply:
Imo `thrust::sort` or `thrust::stable_sort_by_key` solution may be more messier as we would need to construct a segment_id vector sort based a zip_iterator of `{segment_id, actual_key_in_the_segment}` but worth benchmarking I guess.